test/common/ext_authz/ext_authz_impl_test.cc

```diff
@@ -0,0 +1,193 @@
+#include "test/mocks/ssl/mocks.h"
+#include "test/mocks/upstream/mocks.h"
+#include "test/test_common/printers.h"
+#include "test/test_common/utility.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::AtLeast;
+using testing::Invoke;
+using testing::Ref;
+using testing::Return;
+using testing::ReturnRef;
+using testing::WithArg;
+using testing::_;
+
+namespace Envoy {
+namespace ExtAuthz {
+
+class MockRequestCallbacks : public RequestCallbacks {
+public:
+  MOCK_METHOD1(onComplete, void(CheckStatus status));
+};
+
+class ExtAuthzGrpcClientTest : public testing::Test {
+public:
+  ExtAuthzGrpcClientTest()
+      : async_client_(new Grpc::MockAsyncClient()),
+        client_(Grpc::AsyncClientPtr{async_client_}, Optional<std::chrono::milliseconds>()) {}
+
+  Grpc::MockAsyncClient* async_client_;
+  Grpc::MockAsyncRequest async_request_;
+  GrpcClientImpl client_;
+  MockRequestCallbacks request_callbacks_;
+  Tracing::MockSpan span_;
+};
+
+TEST_F(ExtAuthzGrpcClientTest, Basic) {
+  std::unique_ptr<envoy::service::auth::v2::CheckResponse> response;
+
+  {
+    envoy::service::auth::v2::CheckRequest request;
+    Http::HeaderMapImpl headers;
+    EXPECT_CALL(*async_client_, send(_, ProtoEq(request), Ref(client_), _, _))
+        .WillOnce(Invoke([this](const Protobuf::MethodDescriptor& service_method,
+                                const Protobuf::Message&, Grpc::AsyncRequestCallbacks&,
+                                Tracing::Span&,
+                                const Optional<std::chrono::milliseconds>&) -> Grpc::AsyncRequest* {
+          EXPECT_EQ("envoy.service.auth.v2.Authorization", service_method.service()->full_name());
+          EXPECT_EQ("Check", service_method.name());
+          return &async_request_;
+        }));
+
+    client_.check(request_callbacks_, request, Tracing::NullSpan::instance());
+
+    client_.onCreateInitialMetadata(headers);
+    EXPECT_EQ(nullptr, headers.RequestId());
+
+    response.reset(new envoy::service::auth::v2::CheckResponse());
+    ::google::rpc::Status* status = new ::google::rpc::Status();
+    status->set_code(Grpc::Status::GrpcStatus::PermissionDenied);
+    response->set_allocated_status(status);
+    EXPECT_CALL(span_, setTag("ext_authz_status", "ext_authz_unauthorized"));
+    EXPECT_CALL(request_callbacks_, onComplete(CheckStatus::Denied));
+    client_.onSuccess(std::move(response), span_);
+  }
+
+  {
+    envoy::service::auth::v2::CheckRequest request;
+    Http::HeaderMapImpl headers;
+    EXPECT_CALL(*async_client_, send(_, ProtoEq(request), _, _, _))
+        .WillOnce(Return(&async_request_));
+
+    client_.check(request_callbacks_, request, Tracing::NullSpan::instance());
+
+    client_.onCreateInitialMetadata(headers);
+
+    response.reset(new envoy::service::auth::v2::CheckResponse());
+    ::google::rpc::Status* status = new ::google::rpc::Status();
+    status->set_code(Grpc::Status::GrpcStatus::Ok);
+    response->set_allocated_status(status);
```

Comment:
Prefer the `mutable_foo` pattern to raw allocations and `set`. Sorry if I'm being a pain here, it's just a reflection of the Envoy style where we prefer to always have smart pointers to simplify reasoning about correctness. This can be relaxed in tests when really needed, but I don't think this is one of those places.

Reply:
You are fine. It is more a reflection of my old-school C heritage/archaic behavior. The mutable_foo pattern and the smart pointers are much better.